modules/OS/module/src/os_sem_posix.c

```diff
@@ -94,10 +138,84 @@ timespec_init_timeout__(struct timespec* ts, uint64_t us)
     ts->tv_nsec %= 1000000000;
 }
 
-int
-os_sem_take_timeout(os_sem_t sem, uint64_t usecs)
+static int
+os_sem_take_timeout_efd__(os_sem_t sem, uint64_t usecs)
 {
-    VALIDATE(sem);
+    AIM_TRUE_OR_DIE(USES_EFD(sem));
+
+    /** poll() with timeout (in ms) */
+    struct pollfd fds;
+
+    fds.fd = sem->efd;
+    fds.events = POLLIN;
+    fds.revents = 0;
+
+    uint64_t t_start = os_time_monotonic();
+    uint64_t t_poll_started = t_start;
+    int timeout_ms;
+
+    if(usecs == 0) {
+        /* Infinite timeout value when calling poll() */
+        timeout_ms = -1;
+    }
+    else {
+        timeout_ms = usecs / 1000;
+    }
+
+    for(;;) {
+
+        int rv = poll(&fds, 1, timeout_ms);
+
+        if(rv == 0) {
+            /* Timed out */
+            return -1;
+        }
+        if(rv == 1) {
+            if(fds.revents & POLLIN) {
+                /* Descriptor Ready */
+                uint64_t v;
+                if(eventfd_read(sem->efd, &v) == 0) {
+                    /* Acquired. */
+                    return 0;
+                }
+                else {
+                    /**
+                     * Acquired by someone else.
+                     * Retry handled along with EINTR below.
+                     */
+                }
+            }
+        }
+        if(rv == 1 || errno == EINTR) {
+            /*
+             * Calculate remaining timeout (if necessary) and retry.
+             */
+            if(timeout_ms != -1) {
+                uint64_t now = os_time_monotonic();
+                if( (now - t_start) >= usecs ) {
+                    /* Total timeout has elapsed */
+                    return -1;
+                }
+                else {
+                    /* Remaining time to wait. */
+                    timeout_ms -= (now - t_poll_started) / 1000;
```

Comment:
Suggest changing this to `timeout_ms = (now + usecs - t_start)/1000` for consistency with the `(now - t_start) >= usecs` check.

While this scenario is unlikely, if the semaphore is woken up several times the rounding from usec to msec will cause the timeout to be slightly longer than it should be.

I also suggest doing a MIN with 1 so that it doesn't go into a busy-wait loop if it wakes up with less than one millisecond left (or maybe add 999 before dividing by 1000).